programs/drift-competitions/src/state/competition.rs

```diff
@@ -1,32 +1,404 @@
+
+#[derive(Clone, Copy, BorshSerialize, BorshDeserialize, PartialOrd, Ord, PartialEq, Eq, Debug)]
+pub enum CompetitionRoundStatus {
+    Active = 1,
+    PrizeDrawComplete = 2,
+    PrizeAmountComplete = 3,
+    WinnerDrawComplete = 4,
+    WinnerSettlementComplete = 5,
+    Expired = 6,
+}
+
+impl Default for CompetitionRoundStatus {
+    fn default() -> Self {
+        CompetitionRoundStatus::Active
+    }
+}
+
+#[derive(Default, Clone, Copy, AnchorSerialize, AnchorDeserialize, Eq, PartialEq, Debug)]
+pub struct SponsorInfo {
+    pub sponsor: Pubkey,
+    pub min_sponsor_amount: u64,   // always leave this amount for sponsor
+    pub max_sponsor_fraction: u64, // only take this percent of gain above the min amount
+}
+
 #[assert_no_slop]
 #[account(zero_copy)]
 #[derive(Default, Eq, PartialEq, Debug)]
 #[repr(C)]
 pub struct Competition {
     pub name: [u8; 32],
-    pub sponsor: Pubkey,
-
-    pub round_number: u64,
+    pub sponsor_info: SponsorInfo,
 
     // entries
     pub number_of_competitors: u128,
-    pub max_entries_per_competitor: u128, // set a max entry per competitior
     pub number_of_competitors_settled: u128, //starts at zero and you have to settle everyone to know the winner
     pub total_score_settled: u128, // sum of all scores (when num users settled == num users)
+    pub max_entries_per_competitor: u128, // set a max entry per competitior
+
+    // giveaway details
+    pub prize_draw: u128,
+    pub prize_amount: u128,
+    pub prize_base: u128,
+    pub winning_draw: u128,
 
     // scheduling variables
+    pub round_number: u64,
     pub first_round_expiry_ts: i64,
     pub competition_expiry_ts: i64, // when competition ends, perpetual when == 0
     pub round_duration: u64,
+
+    pub status: CompetitionRoundStatus,
+    pub padding: [u8; 7],
 }
 
 impl Size for Competition {
-    const SIZE: usize = 160 + 8;
+    const SIZE: usize = 248 + 8;
 }
 
-const_assert_eq!(Competition::SIZE, std::mem::size_of::<Competition>() + 8);
\ No newline at end of file
+const_assert_eq!(Competition::SIZE, std::mem::size_of::<Competition>() + 8);
+
+impl Size for SponsorInfo {
+    const SIZE: usize = 48 + 8;
+}
+
+const_assert_eq!(SponsorInfo::SIZE, std::mem::size_of::<SponsorInfo>() + 8);
+
+impl Competition {
+    pub fn update_status(&mut self, new_status: CompetitionRoundStatus) -> CompetitionResult {
+        if new_status != CompetitionRoundStatus::Expired {
+            validate!(
+                (new_status as i32) - (self.status as i32 % 5) == 1,
+                ErrorCode::InvalidStatusUpdateDetected,
+                "new status = {:?}, current status = {:?}",
+                new_status,
+                self.status
+            )?;
+            self.status = new_status;
+        }
+
+        Ok(())
+    }
+
+    // calculate the end unix timestamp for round_number N
+    pub fn calculate_round_end_ts(&self) -> DriftResult<i64> {
+        self.first_round_expiry_ts
```

Comment:
feel like sponsor might wanna push back expiry for a round in certain cases, maybe that would be easier if you stored `next_expiry_ts` instead